dist/build.gradle

```diff
@@ -0,0 +1,141 @@
+    provided("commons-httpclient:commons-httpclient:3.0.1")
+    provided("commons-codec:commons-codec:1.4")
+    provided("javax.xml.bind:jaxb-api:2.3.1")
+    optional("org.apache.hive:hive-service:$hiveVersion") {
+        exclude module: "log4j-slf4j-impl"
+    }
+    optional("org.apache.hive:hive-exec:$hiveVersion")
+    optional("org.apache.hive:hive-metastore:$hiveVersion")
+    optional("org.apache.pig:pig:$pigVersion:h2")
+    optional("org.apache.spark:spark-core_${project.ext.scala211MajorVersion}:$spark20Version") {
+        exclude group: 'javax.servlet'
+        exclude group: 'org.apache.hadoop'
+    }
+    optional("org.apache.spark:spark-yarn_${project.ext.scala211MajorVersion}:$spark20Version") {
+        exclude group: 'org.apache.hadoop'
+    }
+    optional("org.apache.spark:spark-sql_${project.ext.scala211MajorVersion}:$spark20Version") {
+        exclude group: 'org.apache.hadoop'
+    }
+    optional("org.apache.spark:spark-streaming_${project.ext.scala211MajorVersion}:$spark20Version") {
+        exclude group: 'org.apache.hadoop'
+    }
+    optional("org.scala-lang:scala-library:$scala211Version")
+    optional("org.scala-lang:scala-reflect:$scala211Version")
+    optional("org.apache.storm:storm-core:$stormVersion") {
+        exclude module: "log4j-slf4j-impl"
+    }
+    provided(project.ext.hadoopClient)
+    provided("org.apache.hadoop:hadoop-common:${project.ext.hadoopVersion}")
+    provided("org.apache.hadoop:hadoop-mapreduce-client-core:${project.ext.hadoopVersion}")
+    provided("org.codehaus.jackson:jackson-mapper-asl:${project.ext.jacksonVersion}")
+    provided("org.codehaus.jackson:jackson-core-asl:${project.ext.jacksonVersion}")
+    optional("joda-time:joda-time:$jodaVersion")
+
+    // This will allow us to pull the sources required to build the uberjar's javadoc and source jars
+    // TODO: SO this pulls the generated java code from the spark project, but NOT the scala code.
+    // Scala code is not exported in the sourceElements configurations because it shouldn't be consumed by javadoc
+    // Scala code will eventually need to be consumed by downstream scaladoc tasks in spark variants
+    // It is also needed here, for original source packaging purposes
+    // Maybe we need to expose "java sources", "scala sources", and "all sources" as separate artifact configurations?
+    additionalSources(project(":elasticsearch-hadoop-mr"))
+    additionalSources(project(":elasticsearch-hadoop-hive"))
+    additionalSources(project(":elasticsearch-hadoop-pig"))
+    additionalSources(project(":elasticsearch-spark-20"))
+    additionalSources(project(":elasticsearch-storm"))
+
+    // This will pull in the regular jar, javadoc jar, and source jar for each project
+    dist(project(":elasticsearch-hadoop-mr"))
+    dist(project(":elasticsearch-hadoop-hive"))
+    dist(project(":elasticsearch-hadoop-pig"))
+    dist(project(":elasticsearch-spark-20"))
+    dist(project(":elasticsearch-storm"))
+}
```

Comment:
We have a lot of repeated project dependencies here. Each configuration is pulling in something different:

`embedded` is pulling in just the java jar file from each project without transitive dependencies in order to create the uber jar.
`additionalSources` is pulling in sources meant for javadoc generation
`dist` is pulling in each integration's jar, javadocJar, and sourceJar to be packed up in the zip file for the download page.

We might need to add a different sources configuration to pull in original sources (scala and java) instead of the current state (java and generated java) for the sake of having an accurate sourcesJar.

This seems like a ton of repeated project dependencies. Is this the right way to do things?

Reply:
You could define the set of projects once and then just loop over them adding them to each different configuration.

Reply:
This did end up being better in the long run. I ended up needing to define a couple more configurations to get javadocs and source jars generating correctly due to the mixed java and scala sources.